dags/financial_data_dag.py

```diff
@@ -0,0 +1,135 @@
+"""
+Downloads stock market data from S&P 500 companies and inserts it into CrateDB.
+
+Prerequisites
+-------------
+In CrateDB, the schema to store this data needs to be created once manually.
+See the file setup/financial_data_schema.sql in this repository.
+
+"""
+import datetime
+import math
+import json
+import requests
+from bs4 import BeautifulSoup
+import yfinance as yf
+import pandas as pd
+
+from airflow import DAG
+from airflow.providers.postgres.operators.postgres import PostgresOperator
+from airflow.operators.python import PythonOperator
+
+def get_sp500_ticker_symbols():
+    """Extracts SP500 companies' tickers from the SP500's wikipedia page"""
+
+    # getting the html code from S&P 500 wikipedia page
+    url = "https://en.wikipedia.org/wiki/List_of_S%26P_500_companies"
+    r_html = requests.get(url,timeout=2.5).text
+    soup = BeautifulSoup(r_html, 'html.parser')
+    components_table = soup.find_all(id="constituents")
+
+    # the first section (index=0) in the components table contains
+    # the headers (which are unimportant in this context) therefore,
+    # only data from index=1 on is taken. moreover, each row in the
+    # table is stored in a <tr> division, so we find all of those.
+    # The data is stored in an array, where each element contains
+    # information about a S&P 500 company.
+    data_rows = components_table[0].find("tbody").find_all("tr")[1:]
+    tickers = []
+
+    # extracting the tickers from the data
+    for row, _ in enumerate(data_rows):
+        stock = list(filter(None,data_rows[row].text.split("\n")))
+        symbol = stock[0]
+
+        if symbol.find('.') != -1:
+            symbol = symbol.replace('.', '-')
+
+        tickers.append(symbol)
+
+    return tickers
+
+def download_yfinance_data_function(start_date):
+    """downloads Adjusted Close data from SP500 companies"""
+
+    tickers = get_sp500_ticker_symbols()
+    data = yf.download(tickers, start=start_date)['Adj Close']
+    return data.to_json()
+
+def prepare_data_function(ti):
+    """creates a list of dictionaries with clean data values"""
+
+    # pulling data (as string)
+    string_data = ti.xcom_pull(task_ids='download_data_task')
+
+    # transforming to json
+    json_data = json.loads(string_data)
+
+    # transforming to dataframe for easier manipulation
+    df = pd.DataFrame.from_dict(json_data, orient='index')
+
+    values_dict = []
+
+    for col, _ in enumerate(df.columns):
```

Comment:
`enumerate` here and on `df.index` is appropriate as you use the index in line 78. Instead of ignoring the value part of the tuple I suggest to directly use it.

With this you should be able to remove the array access lines.

Example:
`for col, closing_date in enumerate(df.columns):` should make `closing_date = df.columns[col]` no longer necessary. Same pattern can be applied for `enumerate(df.index)`